docs_src/design/adr/device-service/0012-DeviceService-Filters.md

```diff
@@ -0,0 +1,152 @@
+
+- provide configuration to specify which filter functions to invoke
+- create the filter
+- invoke the filtering functions
+
+At this time, custom filters will not be supported as the custom filters would not be known by the SDK and therefore could not be specified in configuration.  This is consistent with the app functions SDK and filtering.
+
+#### Function Inflection Point
+
+When instructed to "get" new readings, the function [`execReadDeviceResource`](https://github.com/edgexfoundry/device-sdk-go/blob/0bbcb663a9153978e7e9ef8c297d5988e58906d0/internal/handler/command.go) is called which subsequently calls on the the device service driver's `HandleReadCommands` function to get the latest sensor values from the device.  The driver's `HandleReadCommands` returns the sensor reading data (via array of CommandValues) to be put into an event.
+
+After receiving the `CommandValues` the `execReadDeviceResource` function calls the `cvsToEvent` function convert the `CommandValues` into and `Event/Reading` objects (from `go-mod-core-contracts`) and returns this model to be sent via REST to core data by the rest of the SDK.
+
+It is precisely after the convert to `Event/Reading` objects and before returning that result in `common.SendEvent` (in utils.go) function that the device service should invoke the required filter functions.
+
+Events containing binary values (event.HasBinaryValue), will not be filtered.  Future releases may include binary value filters.
+
+!!! TODO
+    *Where and how would this work for Async events? Need help from the DS team.*
+
+    *Are the Go and C SDKs similar enough to use this design?*
+
+    *Do we care about function order?*
+
+#### Setting Filter Function and Configuration
+
+When filter functions are shared (or appear to be doing the same type of work) between SDKs, the configuration of the similar filter functions should also look similar.  The app functions SDK configuration model for filters should therefore be followed.
+
+While device services do not have pipelines, the inclusion and configuration of filters for device services should have a similar look (to provide symmetry with app services). The configuration has to provide the functions required and parameters to make the functions work - even though the association to a pipeline is not required.  Below is the common app service configuration as it relates to filters:
+
+``` toml
+[Writable.Pipeline]
+    ExecutionOrder = "FilterByDeviceName, TransformToXML, SetOutputData"
+    [Writable.Pipeline.Functions.FilterByDeviceName]
+    [Writable.Pipeline.Functions.FilterByDeviceName.Parameters]
+        DeviceNames = "Random-Float-Device,Random-Integer-Device"
+        FilterOut = "false"
+```
+
+Suggested and hypothetical configuration for the device service reading filters should look something like that below.
+
+``` toml
+[Writable.Filters]
+    ExecutionOrder = "FilterByValueInRange, FilterByResourceNamesMatch"
```

Comment:
Wouldn't you always want to filter by device resource name first? Otherwise you're applying filters to device resources that might not make sense? Do we really want generic value filtering? I would think instead you'd want each filter to apply to a specific device resource?